pyls/rpc_manager.py

```diff
@@ -0,0 +1,171 @@
+        """Send a JSONRPC message with an expected response.
+
+        Args:
+            method (str): The method name of the message to send
+            params (dict): The payload of the message
+
+        Returns:
+            Future that will resolve once a response has been recieved
+
+        """
+        log.debug('Calling %s %s', method, params)
+        request = JSONRPC20Request(_id=uuid1().int, method=method, params=params)
+        request_future = Future()
+        self._sent_requests[request._id] = request_future
+        self._message_manager.write_message(request.data)
+        return request_future
+
+    def notify(self, method, params=None):
+        """Send a JSONRPC notification.
+
+         Args:
+             method (str): The method name of the notification to send
+             params (dict): The payload of the notification
+         """
+        log.debug('Notify %s %s', method, params)
+        notification = JSONRPC20Request(method=method, params=params)
+        self._message_manager.write_message(notification.data)
+
+    def cancel(self, request_id):
+        """Cancel pending request handler.
+
+        Args:
+            request_id (string | number): The id of the original request
+
+        Note:
+            Request will only be cancelled if it has not begun execution.
+        """
+        log.debug('Cancel request %d', request_id)
+        try:
+            self._received_requests[request_id].cancel()
+        except KeyError:
+            log.error('Received cancel for finished/nonexistent request %d', request_id)
+
+    def consume_requests(self):
+        """ Infinite loop watching for messages from the client."""
+        for message in self._message_manager.get_messages():
+            if isinstance(message, JSONRPCBaseResponse):
+                self._handle_response(message)
+            else:
+                self._handle_request(message)
+
+    def _handle_request(self, request):
+        """Execute corresponding handler for the recieved request
+
+        Args:
+            request (JSONRPCBaseRequest): request to act upon
+
+        Note:
+            requests are handled asynchronously if the handler returns a callable, otherwise they are handle
+            synchronously by the main thread
+        """
+        if self._shutdown and request.method != 'exit':
+            return
+
+        params = request.params if request.params is not None else {}
+        try:
+            maybe_handler = self._message_handler(request.method, params)
+        except KeyError:
+            log.debug("No handler found for %s", request.method)
+            self._message_manager.write_message(
+                JSONRPC20Response(_id=request._id, error=JSONRPCMethodNotFound()._data).data)
+            return
+
+        if request._id in self._received_requests:
+            log.error('Received request %s with duplicate id', request.data)
+        elif callable(maybe_handler):
```

Comment:
I think this might be the wrong place to check callable. How can I have two linting hooks, one sync and one async right now?

Reply:
Linting is a notification so I don't think it wouldn't be effected by this. Perhaps i didn't mention this, but the plan is to have the server do the resolution around having some async and some non-async hooks. i.e. for each handler there would be some merge logic to determine whether to return an aggregate result or a callable